### ISSUE TITLE
istioctl experimental upgrade does not KubernetesResourcesSpec (k8s) related changes

### ISSUE THREAD

**Issue by paweloczadly** [istio/istio#20066]
**Bug description**

`istioctl experimental upgrade` does not display `KubernetesResourcesSpec` (k8s) related changes.

Example configuration:

```yaml
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  trafficManagement:
    enabled: true
    components:
      pilot:
        k8s:
          priorityClassName: istio-high-prio
  policy:
    enabled: false
  security:
    enabled: true
  autoInjection:
    enabled: true
  configManagement:
    enabled: true
  telemetry:
    enabled: true
    components:
      telemetry:
        k8s:
          env:
            - name: TEST_VAR
              value: testvalue

  values:
    global:
      disablePolicyChecks: false
      controlPlaneSecurityEnabled: false
      proxy:
        accessLogFile: /dev/stdout
        resources:
          requests:
            cpu: 10m
            memory: 40Mi

    pilot:
      autoscaleEnabled: false

    gateways:
      istio-egressgateway:
        autoscaleEnabled: false
      istio-ingressgateway:
        autoscaleEnabled: false

    prometheus:
      enabled: true

    kiali:
      enabled: true
      createDemoSecret: true
```

Current output:

```
$ istioctl experimental upgrade -f minikube-istiocontrolplane.yaml
Client - istioctl version: 1.4.3

Upgrade - target version: 1.4.3

Control Plane - citadel pod - istio-citadel-6d6cbfdddb-d566r - version: 1.4.3
Control Plane - galley pod - istio-galley-f65ffd68d-2mmf4 - version: 1.4.3
Control Plane - ingressgateway pod - istio-ingressgateway-5c97dd657f-p2kbh - version: 1.4.3
Control Plane - pilot pod - istio-pilot-d77b658ff-8cmqn - version: 1.4.3
Control Plane - sidecar-injector pod - istio-sidecar-injector-75ff97b4d8-54sgb - version: 1.4.3
Control Plane - telemetry pod - istio-telemetry-6cc6cf7b64-kwv6z - version: 1.4.3

Upgrade version check passed: 1.4.3 -> 1.4.3.

Upgrade check: Warning!!! The following values will be changed as part of upgrade. If you have not overridden these values, they will change in your cluster. Please double check they are correct:
sidecarInjectorWebhook:
  replicaCount: 1 ->
  rollingMaxSurge: 100% ->
  rollingMaxUnavailable: 25% ->

Confirm to proceed [y/N]?
```

When the changes are applied `pilot` and `telemetry` deployments are updated.

- [ ] Configuration Infrastructure
- [ ] Docs
- [x] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [x] User Experience
- [ ] Developer Infrastructure

**Expected behavior**

Changes in pilot and telemetry are displayed.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

```
$ istioctl version --remote
client version: 1.4.3
control plane version: 1.4.3
data plane version: 1.4.3 (59 proxies), 1.4.2 (14 proxies)
```

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-13T11:51:44Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-gke.2", GitCommit:"0f206d1d3e361e1bfe7911e1e1c686bc9a1e0aa5", GitTreeState:"clean", BuildDate:"2019-11-25T19:35:58Z", GoVersion:"go1.12.12b4", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**

istioctl

**Environment where bug was observed (cloud vendor, OS, etc)**

* GKE
* Mac OS X
* istioctl installed from  brew

**Comment by elfinhe** [comment]
thanks @paweloczadly, I will take a look next week.

**Comment by paweloczadly** [comment]
Also, `rollingMaxSurge` and `rollingMaxUnavailable` are displayed in every upgrade execution and it's not possible to set them using `spec.values.sidecarInjectorWebhook.rollingMaxSurge` and `spec.values.sidecarInjectorWebhook.rollingMaxUnavailable` as described [here](https://istio.io/docs/reference/config/installation-options/#sidecarinjectorwebhook-options). 

Once they are set, `experimental upgrade` displays the following errors:

```
Client - istioctl version: 1.4.3

Error: failed to generate values from file: minikube-istiocontrolplane.yaml, error: input file failed validation with the following errors: unknown field "rollingMaxUnavailable" in v1alpha1.SidecarInjectorConfig

...
```

and

```
Client - istioctl version: 1.4.3

Error: failed to generate values from file: minikube-istiocontrolplane.yaml, error: input file failed validation with the following errors: unknown field "rollingMaxSurge" in v1alpha1.SidecarInjectorConfig
```

**Comment by elfinhe** [comment]
The k8s issue is a duplicate of https://github.com/istio/istio/issues/19135, which will be fixed in 1.5.0.

**Comment by elfinhe** [comment]
As for the sidecarInjector issue, you might need to use  IstioControlPlaneSpec API instead of the global `values`.

For example:
`istioctl manifest apply --set autoInjection.components.injector.k8s.strategy.rollingUpdate.maxSurge="100%"`

or in yaml format:

```yaml
  # Auto injection feature
  autoInjection:
    enabled: true
    components:
      injector:
        enabled: true
        k8s:
          replicaCount: 1
          strategy:
            rollingUpdate:
              maxSurge: "100%"
              maxUnavailable: "25%"
```

**Comment by paweloczadly** [comment]
Thanks for response @elfinhe. It seems experimental upgrade also doesn't respect the YAML config which you pasted above:

```
istioctl experimental upgrade -f minikube-istiocontrolplane.yaml
Client - istioctl version: 1.4.3

Upgrade - target version: 1.4.3

Control Plane - citadel pod - istio-citadel-748fbf75d-7v5hp - version: 1.4.3
Control Plane - galley pod - istio-galley-66b8db756-h6dcj - version: 1.4.3
Control Plane - pilot pod - istio-pilot-d77b658ff-vcph7 - version: 1.4.3
Control Plane - sidecar-injector pod - istio-sidecar-injector-58f966587b-b6mgt - version: 1.4.3
Control Plane - telemetry pod - istio-telemetry-6b6cf8d6d8-trbfs - version: 1.4.3

Upgrade version check passed: 1.4.3 -> 1.4.3.

Upgrade check: Warning!!! The following values will be changed as part of upgrade. If you have not overridden these values, they will change in your cluster. Please double check they are correct:
sidecarInjectorWebhook:
  replicaCount: 1 ->
  rollingMaxSurge: 100% ->
  rollingMaxUnavailable: 25% ->

Confirm to proceed [y/N]?
```

And here is my whole configuration:

```
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  tag: 1.4.3
  trafficManagement:
    enabled: true
    components:
      pilot:
        k8s:
          priorityClassName: istio-high-prio
          hpaSpec:
            minReplicas: 1
            maxReplicas: 1
  policy:
    enabled: false
  security:
    enabled: true
    components:
      citadel:
        k8s:
          priorityClassName: istio-high-prio
  autoInjection:
    enabled: true
    components:
      injector:
        k8s:
          priorityClassName: istio-high-prio
          replicaCount: 1
          strategy:
            rollingUpdate:
              maxSurge: "100%"
              maxUnavailable: "25%"
  configManagement:
    enabled: true
    components:
      galley:
        k8s:
          priorityClassName: istio-high-prio
  telemetry:
    enabled: true
    components:
      telemetry:
        k8s:
          priorityClassName: istio-high-prio
          hpaSpec:
            minReplicas: 1
            maxReplicas: 1
  gateways:
    components:
      ingressGateway:
        enabled: false

  values:
    global:
      disablePolicyChecks: false
      controlPlaneSecurityEnabled: false
      proxy:
        accessLogFile: /dev/stdout
        resources:
          requests:
            cpu: 10m
            memory: 40Mi

    kiali:
      enabled: true
      createDemoSecret: true
```

**Comment by elfinhe** [comment]
Yes, you are right. That will also be fixed in be fixed in 1.5.0 by istio/operator#671. The upgrade will still use your yaml in this case, but will not compare the k8s specs in components in the upgrade check.
